tr2/savegame: migrate to BSON save format

lahm86 opened the pull request:
Resolves #2662.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This is a step towards BSON saves, initially tackling making the original logic more like TR1X. We still have a hybrid of new modules and the decomp/savegame module; I'll build on this PR incrementally with the goal being BSON plus legacy properly segmented.

I'd appreciate some initial testing of what's done so far, that is results in game should be the same for save/load, stats, and passport/slot display/counters.

lahm86 committed:
tr2/types: rename START_INFO to RESUME_INFO

This renames the START_INFO struct to RESUME_INFO, along with the
relevant variables that use it.

lahm86 committed:
tr2/misc: eliminate current stats

This removes current_stats in favour of using the current info for
levels.

lahm86 committed:
tr2/savegame: hide legacy functions

This updates callers to use the Savgame_Save and Savegame_Load
functions in order to hide the current legacy approach. TR1 also had
duplicated function declarations, so these are removed.

lahm86 committed:
tr2/savegame: move resume info to module storage

This moves the resume info array into the common savegame module.

lahm86 committed:
tr2/savegame: refactor savegame logic

This is a general refactor of the savegame logic, primarily to bring it
closer to TR1 in preparation for introducing strategies.

rr- reviewed (approved):
From the quick tests everything seems to be in order. I haven't tested end game stats. The code also LGTM.

aredfan commented:
> I'd appreciate some initial testing of what's done so far, that is results in game should be the same for save/load, stats, and passport/slot display/counters.

LGTM 👍

lahm86 committed:
game: move game mode flag to game module

This makes the bonus flag common between both games, now accessible in
the game module. This paves the way for TR2 to offer alternate game
modes in the future (beyond current scope).

lahm86 committed:
tr2/savegame: encapsulate save IO

This removes the disjointed OG approach of saving and loading by
storing some variables globally before later interperting them, in
favour of using the same approach as TR1's legacy saves, where the
buffer is read/written in one move.

lahm86 committed:
tr2/savegame: move pre/post processing logic

This moves the pre and post processing logic to the main savegame
module.

lahm86 committed:
tr2/savegame: isolate legacy savegame module

This isolates TR2's legacy savegame logic into its own module, similar
to TR1, and updates related usage throughout.

lahm86 commented:
The latest commits involve a lot of moving code around, but we're very close to being able to start the BSON module 😀 I'd appreciate another general save/load/stats test for TR2. It'd also be good to verify NG+ works, and in addition, for TR1 as well because I've moved this from being a global to a module var, common to both games. Thanks!

aredfan commented:
TR2X looks good including the NG+ flag. The NG+ saves for TR1X works as expected too. 👍

lahm86 committed:
tr2/savegame: introduce strategy concept

This introduces the concept of savegame strategies, like TR1X, with the
legacy strategy being the only one currently.

lahm86 committed:
tr2/savegame: implement BSON saves

This implements BSON save games, similar to TR1, and opts for it to be
used in place of legacy for writing. Saves will also be stored in a
separate saves directory.

lahm86 committed:
savegame: move common types

This moves the savegame info and format types to TRX. More intricate
work will be needed in TR1 to allow moving more, so that is beyond
current scope.

lahm86 commented:
BSON saves are in, and I've rebased against develop to get everything up to date. I unfortunately hit a problem with skidoo drivers  - because of the way their object IDs are swapped, I've had to revert to calling `obj->handle_save_func` within the strategy rather than in the common module, otherwise the previous object ID is lost.

I'll continue with testing myself but wanted to mark this as ready to get more eyes looking for any issues 👀. Things to look out for:

- Carried items
- Lift
- Boat
- Skidoos (drivers and black skidoos should work too currently)
- Thrown flares
- One shot music
- NG+
- Angry monks
- Stats
- Inventory
- Flipmaps
- Golden Mask
- And anything else you can think of! Animation state, gun status etc

There is also a temporary uncompressed save file stored for debugging any issues we hit - it's `debug.json` in the saves folder (only the latest save). I will squash out the code for this before we finalise everything. Thanks!

aredfan commented:
Issues I noticed (this list may grow so keep an eye on it):

- [x] Lara doesn't stay on fire - `/play 1` `/tp 65`

I'll hijack this comment to add things I find too (Lahm):
- [x] - `/play 12`, `/give all`, save/load (item swapping foulplay)
- [x] Equip Shotgun, M16, Harpoon or GL; start firing; save; load a _different_ level, then reload the first save (likely crash)

lahm86 commented:
@aredfan Thanks, updates pushed. Note that I've changed the format slightly so you'll need to delete your current (BSON) saves. I think we can afford to do this until we get it all working, rather than putting in extra code.

rr- commented on data/tr2/ship/cfg/TR2X_gameflow_level.json5:
```diff
@@ -3,6 +3,7 @@
     "main_menu_picture": "data/images/title_eu.png",
     "savegame_fmt_legacy": "savegame.%d",
+    "savegame_fmt_bson": "save_tr2_%02d.dat",
 
```
These collide with the OG saves, can we make the pattern separate?

rr- commented on docs/tr2/README.md:
```diff
@@ -310,6 +310,7 @@ as Notepad.
 - added Linux builds
 - added macOS builds
 - added .jpeg/.png screenshots
+- added BSON savegame support
```
I think the meaning can be vague to the average recipient, instead I'd focus on the benefits for the TRLE and Rando communities both here and in the changelog.

rr- commented on src/tr2/game/savegame/common.c:
```diff
@@ -42,25 +194,441 @@ int32_t Savegame_GetSlotCount(void)
+        char *full_path = String_Format("%s/%s", SAVES_DIR, file_name);
+        MYFILE *const fp = File_Open(full_path, FILE_OPEN_WRITE);
+        if (fp != nullptr) {
+            m_SaveCounter++;
```
Let's move this to the if outside the for loop as theoretically it can increment this counter once per each successful saving strategy.

lahm86 replied:
Done, I've had to add a decrement check at the end too if the result isn't true - the idea is the strategies access it with `Savegame_GetCounter` so we need to have incremented it beforehand.

rr- commented on src/tr2/game/game_flow/sequencer_events.c:
```diff
@@ -114,23 +115,29 @@ static DECLARE_GF_EVENT_HANDLER(M_HandlePlayLevel)
+        if (!Savegame_Load(slot_num)) {
+            LOG_ERROR("Failed to load save file!");
+            Game_SetCurrentLevel(nullptr);
+            GF_SetCurrentLevel(nullptr);
```
Are these two setters necessary?

lahm86 replied:
Good point, it seems to be fine without it. I can follow up with TR1 (it's the same there) after this PR while doing some refactoring on the SG logic there to do the same.

rr- reviewed (commented):
Thank you. Looks terrific!

rr- reviewed (approved):
Thank you. Code LGTM. I haven't tested the game.

lahm86 commented:
^ Removed the temporary `debug.json` output in last commit.

aredfan reviewed (approved):
LGTM, and well done. 👍
